validator/db/kv/attestation_history_new.go

```diff
@@ -0,0 +1,149 @@
+package kv
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	bolt "go.etcd.io/bbolt"
+	"go.opencensus.io/trace"
+)
+
+const (
+	// The size of each data entry in bytes for the source epoch (8 bytes) and signing root (32 bytes).
+	uint64Size             = 8
+	latestEpochWrittenSize = uint64Size
+	targetSize             = uint64Size
+	sourceSize             = uint64Size
+	signingRootSize        = 32
+	historySize            = targetSize + sourceSize + signingRootSize
+	minimalSize            = latestEpochWrittenSize
+)
+
+// AttestationHistoryNew stores the historical attestation data needed
+// for protection of validators.
+type AttestationHistoryNew struct {
+	TargetToSource     map[uint64]*HistoryData
+	LatestEpochWritten uint64
+}
+
+// HistoryData stores the needed data to confirm if an attestation is slashable
+// or repeated.
+type HistoryData struct {
+	Source      uint64
+	SigningRoot []byte
+}
+
+// encapsulated history data
+type encHistoryData []byte
+
+func (hd encHistoryData) assertSize() error { // pointer receiver will also work
+	if len(hd) == 0 || len(hd) < minimalSize || (len(hd)-minimalSize)%historySize != 0 {
+		return fmt.Errorf("encapsulated data size: %d is not a multiple of entry size: %d", len(hd), historySize)
+	}
+	return nil
+}
+
+func newAttestationHistoryArray(target uint64) encHistoryData {
+	enc := make(encHistoryData, latestEpochWrittenSize+(target%params.BeaconConfig().WeakSubjectivityPeriod)*historySize+historySize)
+	return enc
+}
+
+func getLatestEpochWritten(ctx context.Context, data encHistoryData) (uint64, error) {
```

Comment:
Can this be changed into `encHistoryData.getLatestEpochWritten()`? My main reason for defining a new type was so that this and similar funcs are grouped/scoped inside that type.

Reply:
thanks